selfdrive/updated.py

```diff
@@ -1,57 +1,267 @@
+  os.umask(0o077)
+  os.mkdir(FINALIZED)
+  os.chdir(OVERLAY_MERGED)
+  for root, dirs, files in os.walk('.', topdown=True):
+    for obj_name in dirs:
+      dup_ovfs_object(inode_map, os.path.join(root, obj_name), FINALIZED)
+    for obj_name in files:
+      dup_ovfs_object(inode_map, os.path.join(root, obj_name), FINALIZED)
+  os.chdir(BASEDIR)  # otherwise umount will fail later with "target is busy"
+
+def attempt_update():
+  upd_lock_fd = open(UPDATE_LOCK, 'r')
+  try:
+    fcntl.flock(upd_lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+  except OSError:
+    print("couldn't get update lock; is another updated running? going back to sleep...")
+    upd_lock_fd.close()
+    return
+
+  cloudlog.info("attempting git update inside staging overlay")
+  os.chdir(OVERLAY_MERGED)
+
+  # Un-set the successful flag to prevent the finalized tree from being
+  # activated later if the update fails in some way.
+  if os.path.isfile(os.path.join(FINALIZED, ".update_succeeded")):
+    os.remove(os.path.join(FINALIZED, ".update_succeeded"))
+  os.system("sync")
+
+  r = run(NICE_LOW_PRIORITY + ["git", "fetch"], OVERLAY_MERGED)
+  cloudlog.info("git fetch success: %s", r)
+
+  cur_hash = run(["git", "rev-parse", "HEAD"], OVERLAY_MERGED).rstrip()
+  upstream_hash = run(["git", "rev-parse", "@{u}"], OVERLAY_MERGED).rstrip()
+
+  cloudlog.info("comparing %s to %s" % (cur_hash, upstream_hash))
+  if cur_hash != upstream_hash:
+    cloudlog.info("git reset in progress")
+    r = [run(NICE_LOW_PRIORITY + ["git", "reset", "--hard", "@{u}"], OVERLAY_MERGED),
+         # TODO: with scons, is clean still required?
+         run(NICE_LOW_PRIORITY + ["git", "clean", "-xdf"], OVERLAY_MERGED),
+         run(NICE_LOW_PRIORITY + ["git", "submodule", "init"], OVERLAY_MERGED),
+         run(NICE_LOW_PRIORITY + ["git", "submodule", "update"], OVERLAY_MERGED)]
+    cloudlog.info("git reset success: %s", '\n'.join(r))
+
+    cloudlog.info("update successful!")
+  else:
+    cloudlog.info("nothing new from git at this time")
+
+  # Make sure the update succeeded flag lands on disk LAST, only when the
+  # local git repo and OP install state are in a consistent state.
+  os.system("sync")
+  Path(os.path.join(OVERLAY_MERGED, ".update_succeeded")).touch()
+  os.system("sync")
+
+  fcntl.flock(upd_lock_fd, fcntl.LOCK_UN)
+  upd_lock_fd.close()
+
+def update_params(with_date=False):
+  params = Params()
+  if os.path.isfile(os.path.join(FINALIZED, ".update_succeeded")):
     try:
-      r = subprocess.check_output(["git", "--no-pager", "show", "@{u}:RELEASES.md"])
+      # TODO: This shouldn't be binary
+      with open(os.path.join(FINALIZED, "RELEASES.md"), "rb") as f:
+        r = f.read()
       r = r[:r.find(b'\n\n')] # Slice latest release notes
       params.put("ReleaseNotes", r + b"\n")
-    except:
+    except Exception:
       params.put("ReleaseNotes", "")
+    params.put("UpdateAvailable", "1")
+  else:
+    params.put("UpdateAvailable", "0")
 
+  if with_date:
```

Comment:
This should probably moved inside the `.update_succeeded` check.

Reply:
Personal preference, I would rename the parameter to something like `write_last_update_time`.